### PR TITLE
Fix incorrect references to "HttpClientFactory"

### LINKED ISSUE #15879: If the problem is the HttpMessageHandler, why not make it more clear?
&gt; This factory assigns the `HttpClient` an `HttpMessageHandler` from a pool it manages.

Then the `HttpClient` can be disposed, right?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet-architecture**
* Technology: **microservices**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### LINKED ISSUE #16422: This class doesn't exist
There is an interface called IHttpClientFactory and a DefaultHttpClientFactory but as far as I can tell there is no concrete class called HttpClientFactory. If I'm wrong please point to the actual NuGet/repo where this class exists. I've looked high and low for it. 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#feedback)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet-architecture**
* Technology: **microservices**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### PR DESCRIPTION
**Pull request by mvelosop**
## Summary

- Revise main article about the "HttpClientFactory"
- Fix references

Fixes https://github.com/dotnet/docs/issues/16422
Fixes https://github.com/dotnet/docs/issues/15879

### CODE REVIEW

**Comment by nishanil** [comment]
@mvelosop resolve conflicts?

**Comment by mvelosop** [comment]
Done @nishanil

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 1]
```diff
@@ -1,62 +1,64 @@
...
+But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object in long-running processes, as could be some sort of server. In this case, a singleton or static `HttpClient` doesn't handle DNS changes, as explained in this [issue](https://github.com/dotnet/corefx/issues/11224) at the dotnet/corefx GitHub repository.
 
-To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced a new `HttpClientFactory` that can also be used to implement resilient HTTP calls by integrating Polly with it.
+To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced the <xref:System.Net.Http.IHttpClientFactory> interface that's implemented by the `DefaultHttpClientFactory`, that can't be instantiated directly, that can also be used to implement resilient HTTP calls by integrating Polly with it.
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
To address the issues mentioned above and to make `HttpClient` instances manageable, .NET Core 2.1 introduced the <xref:System.Net.Http.IHttpClientFactory> interface which can be used to configure and create HttpClient instances in an app. It also provides extensions for Polly-based middleware to take advantage of delegating handlers in HttpClient.
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 2]
```diff
@@ -1,62 +1,64 @@
...
 
-[Polly](http://www.thepollyproject.org/) is transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.
+[Polly](http://www.thepollyproject.org/) is a transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.
 
-## What is HttpClientFactory
+## What is the IHttpClientFactory
 
-`HttpClientFactory` is designed to:
+<xref:System.Net.Http.IHttpClientFactory> is an interface with the single method `CreateClient()` that should return an <xref:System.Net.Http.HttpClient>.
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 3]
```diff
@@ -1,62 +1,64 @@
...
 
-- Provide a central location for naming and configuring logical `HttpClient` objects. For example, you may configure a client (Service Agent) that's pre-configured to access a specific microservice.
+The current implementation is `DefaultHttpClientFactory`, that also implements <xref:ystem.Net.Http.IHttpMessageHandlerFactory>, with the following features:
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
`IHttpClientFactory` offers the following benefits:
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 4]
```diff
@@ -1,62 +1,64 @@
...
+> The implementation of `IHttpClientFactory` (`DefaultHttpClientFactory`) is tightly tied to the dependency injection (DI) implementation in the `Microsoft.Extensions.DependencyInjection` NuGet package. For more information about using other dependency injection containers, see this [GitHub discussion](https://github.com/dotnet/extensions/issues/1345).
 
-## Multiple ways to use HttpClientFactory
+## Multiple ways to use IHttpClientFactory
 
-There are several ways that you can use `HttpClientFactory` in your application:
+There are several ways that you can use `IHttpClientFactory` in your application:
 
-- Use `HttpClientFactory` Directly
+- Use `IHttpClientFactory` Directly
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
- Basic usage
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 5]
```diff
@@ -1,62 +1,64 @@
...
 
-To configure the above structure, add `HttpClientFactory` in your application by installing the `Microsoft.Extensions.Http` NuGet package that includes the `AddHttpClient()` extension method for `IServiceCollection`. This extension method registers the `DefaultHttpClientFactory` to be used as a singleton for the interface `IHttpClientFactory`. It defines a transient configuration for the `HttpMessageHandlerBuilder`. This message handler (`HttpMessageHandler` object), taken from a pool, is used by the `HttpClient` returned from the factory.
+To configure the above structure, install the [`Microsoft.Extensions.Http`](https://www.nuget.org/packages/Microsoft.Extensions.Http/) NuGet package that includes the `AddHttpClient()` extension method for `IServiceCollection`. This extension method registers the `DefaultHttpClientFactory` as a singleton, as the implementation of `IHttpClientFactory`. It defines a transient configuration for the <xref:Microsoft.Extensions.Http.HttpMessageHandlerBuilder>. This message handler (<xref:System.Net.Http.HttpMessageHandler> object), taken from a pool, is used by the `HttpClient` returned from the factory.
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: revert this changes. previous one reads better.

File: docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md [hunk 6]
```diff
@@ -1,22 +1,22 @@
 ---
 title: Implement HTTP call retries with exponential backoff with Polly
-description: Learn how to handle HTTP failures with Polly and HttpClientFactory.
+description: Learn how to handle HTTP failures with Polly and IHttpClientFactory.
 ms.date: 01/30/2020
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md]: update ms date

File: docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md [hunk 7]
```diff
@@ -1,22 +1,22 @@
...
+However, writing your own custom code to use Polly’s library with HttpClient can be significantly complex. In the original version of eShopOnContainers, there was a [ResilientHttpClient building-block](https://github.com/dotnet-architecture/eShopOnContainers/commit/0c317d56f3c8937f6823cf1b45f5683397274815#diff-e6532e623eb606a0f8568663403e3a10) based on Polly. But with the release of [IHttpClientFactory](use-httpclientfactory-to-implement-resilient-http-requests.md), implementing resilient HTTP communication with Polly has become much simpler, so that building-block was deprecated from eShopOnContainers.
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md]: We should remove this para altogether. We don't have to really mention about a historic implementation that does not add any value in this context.

File: docs/architecture/microservices/implement-resilient-applications/implement-circuit-breaker-pattern.md [hunk 8]
```diff
@@ -20,11 +20,11 @@ Therefore, you need some kind of defense barrier so that excessive requests stop
 
 The Circuit Breaker pattern has a different purpose than the "Retry pattern". The "Retry pattern" enables an application to retry an operation in the expectation that the operation will eventually succeed. The Circuit Breaker pattern prevents an application from performing an operation that's likely to fail. An application can combine these two patterns. However, the retry logic should be sensitive to any exception returned by the circuit breaker, and it should abandon retry attempts if the circuit breaker indicates that a fault is not transient.
 
-## Implement Circuit Breaker pattern with HttpClientFactory and Polly
+## Implement Circuit Breaker pattern with `IHttpClientFactory` and Polly
```
Comment (nishanil) [inline docs/architecture/microservices/implement-resilient-applications/implement-circuit-breaker-pattern.md]: Fix ms date in this file.

**Review by nishanil — COMMENTED** [review]
@mvelosop Left a few comments for you to address.

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md (hunk 2 above)
Comment (nishanil) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: Remove this sentence as it is only describing what the interface is.

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 9]
```diff
@@ -1,62 +1,64 @@
...
 
-To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced a new `HttpClientFactory` that can also be used to implement resilient HTTP calls by integrating Polly with it.
+To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced the <xref:System.Net.Http.IHttpClientFactory> interface that's implemented by the `DefaultHttpClientFactory`, that can't be instantiated directly, that can also be used to implement resilient HTTP calls by integrating Polly with it.
 
-[Polly](http://www.thepollyproject.org/) is transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.
+[Polly](http://www.thepollyproject.org/) is a transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.
 
-## What is HttpClientFactory
+## What is the IHttpClientFactory
```
[the remark this answers is not in the document]
Comment (nishanil) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: IMO, we are talking about the benefits here. The earlier para concludes with what IHttpClientFactory is.

[the remark this answers is not in the document]
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: That's right!

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md (hunk 3 above)
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: Well... not really, it'd be the current implementation of...

Will update!

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md (hunk 5 above)
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: ok

**Comment by mvelosop** [comment]
@nishanil, don't approve/merge this just yet, I'm checking another related issue and might add something.

**Comment by mvelosop** [comment]
Done @nishanil

**Comment by nishanil** [comment]
Conflicts and warnings :(

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 10]
```diff
@@ -1,62 +1,64 @@
...
 
 ## Issues with the original HttpClient class available in .NET Core
 
 The original and well-known <xref:System.Net.Http.HttpClient> class can be easily used, but in some cases, it isn't being properly used by many developers.
 
-As a first issue, while this class is disposable, using it with the `using` statement is not the best choice because even when you dispose `HttpClient` object, the underlying socket is not immediately released and can cause a serious issue named ‘sockets exhaustion’. For more information about this issue, see [You're using HttpClient wrong and it's destabilizing your software](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/) blog post.
+As a first issue, while this class implements `IDisposable`, using it with the `using` statement is not the best choice because when you dispose an `HttpClient` object, the underlying socket is not immediately released and this can cause a serious issue named ‘sockets exhaustion’. For more information about this issue, see the [You're using HttpClient wrong and it's destabilizing your software](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/) blog post.
```
[the remark this answers is not in the document]
Comment (nishanil) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: These changes didn't go in?

[the remark this answers is not in the document]
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: Nope, they didn't.

I still don't get it 🙁. 

I work from the changed files tabs, because that's the only way to batch updates, but it seems like there are different sets of suggestions.

Creating another commit...

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 11]
```diff
@@ -1,62 +1,64 @@
...
 
-But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object. In this case, a singleton or static `HttpClient` doesn't respect DNS changes, as explained in this [issue](https://github.com/dotnet/corefx/issues/11224) at the dotnet/corefx GitHub repository.
+But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object in long-running processes, as could be some sort of server. In this case, a singleton or static `HttpClient` doesn't handle DNS changes, as explained in this [issue](https://github.com/dotnet/corefx/issues/11224) at the dotnet/corefx GitHub repository.
```
[the remark this answers is not in the document]
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]: I'm tweaking this one a bit for you to review. I wouldn't mention the "to improve performance" but focus on the "long-running" processes, b/c that's when the DNS refresh problem could be an issue.

**Review by nishanil — APPROVED** [review]
Looks good @mvelosop !
